Major refactor and optimizations

SHv2 opened the pull request:
Refactored a large portion of the code and optimized portions of code that was being called very frequently.

SHv2 committed:
Refactored code for better readability. Added fps calculation based on detected refresh rate for determining the graph draw speed.

SHv2 committed:
Cleaned up the color selection logic. Removed unnecessary variables.

SHv2 committed:
Don't mix variable names, especially when you're up late putzing around....

SHv2 committed:
Added comments.

SHv2 committed:
Moved to C# 9 to be able to pass pointers to functions.

SHv2 committed:
First pass at optimizing the color selection.

SHv2 committed:
Cleaned up the color logic and removed combinations that aren't possible.

SHv2 committed:
Removed the creation of an unnecessary Createcircle object. Flipped a loop to a do/while to prevent a pointless initial check. Fixed a color missing from the energy mode.

SHv2 committed:
Removed from the Energy mode color that didn't conform to the README.

SHv2 committed:
Bumped version.

SHv2 committed:
Merging in update to C# 9 and color selection optimizations.

SHv2 committed:
Update AssemblyInfo to match version.

SHv2 committed:
Add a delay to the destruction of the panel to pause the garbage collection event. Without the slight delay on levels with 1,000s of notes, there is a noticeable pause after a level completes and goes back to the main menu.

SHv2 committed:
Fixed incorrect first param

SHv2 committed:
Minor nit. Inlined the colorOverride and sideColor lines.

MCJack123 commented on PerformanceMeterController.cs:
```diff
@@ -77,20 +80,6 @@ public class PerformanceMeterController : MonoBehaviour {
             text.GetComponent<RectTransform>().localPosition = new Vector3(-0.3f, -0.25f, 0.1f);
             text.GetComponent<RectTransform>().sizeDelta = new Vector2(75.0f, 25.0f);
 
-            GameObject textObj2 = new GameObject("Label");
```
Any reason this was removed? I'd prefer to keep credits in there, even if it's very small.

MCJack123 commented on Properties/AssemblyInfo.cs:
```diff
@@ -10,7 +10,7 @@
 [assembly: AssemblyCompany("")]
 [assembly: AssemblyProduct("PerformanceMeter")]
-[assembly: AssemblyCopyright("Copyright © 2021 JackMacWindows")]
+[assembly: AssemblyCopyright("Copyright © 2022 SHv2")]
```
This is not your original code, even if you did rewrite/restyle a good chunk of it. Please restore my name - I'll add you to credits in the readme.

MCJack123 commented on PluginConfig.cs:
```diff
@@ -10,25 +10,37 @@
         public virtual float animationDuration { get; set; } = 3.0f;
-        public virtual int color { get; set; } = 0xFF0000;
-        public virtual int secondaryColor { get; set; } = 0x0000FF;
+        public virtual Color color { get; set; } = new Color(1f, 0f, 0f);
```
This will likely break the existing config files. I'll have to check that this is OK and possibly revert this or convert from the form in the file.

MCJack123 commented on PluginConfig.cs:
```diff
@@ -10,25 +10,37 @@
-        public virtual int side { get; set; } = (int)MeasurementSide.Both;
-        public virtual int secondaryMode { get; set; } = (int)MeasurementMode.None;
-        public virtual int secondarySide { get; set; } = (int)MeasurementSide.Both;
+        public virtual MeasurementMode mode { get; set; } = MeasurementMode.Energy;
```
Are you sure these enums can be saved to the file properly? IIRC the reason I used `int`s instead of the raw `enum` was because it would error as it didn't know how to serialize the entry, but maybe newer versions of BSIPA can handle it better.

MCJack123 commented on PluginConfig.cs:
```diff
@@ -10,25 +10,37 @@
-        public virtual int secondarySide { get; set; } = (int)MeasurementSide.Both;
+        public virtual MeasurementMode mode { get; set; } = MeasurementMode.Energy;
+        public virtual MeasurementSide side { get; set; } = MeasurementSide.Both;
+        public virtual Color sideColor { get; set; } = Color.white;
```
I don't quite understand the purpose of the `sideColor` setting. Why is it necessary to have a distinction between this and just the normal `color` override? If the user wants to change the color of each side, they can just change the normal `color` variable.

MCJack123 reviewed (changes requested):
Thank you for taking the time to fix up my crappy code :). Sorry it took so long - I've been very busy for the past month, and I've just gotten some time to look at this properly. I also wanted to wait for the next update to push this out.

This looks pretty good, but I have a few issues with some of your changes, mainly revolving around the config changes. Switching to storing the enums/colors directly as objects will break existing config files, forcing users to fix them back to their preferred settings. I've got a solution using the `IPA.Config.Attributes.UseConverter` attribute that I can merge in, which keeps the file format the same but still allows the code to keep the native types.

MCJack123 committed:
Fixed config conversion, credits adjustments

MCJack123 committed:
Updated for Beat Saber 1.20.0

MCJack123 committed:
Updated version

MCJack123 committed:
Fixed side colors to be constant-ish, fixed issues when no data

Making side colors constant fixes an issue when carrying over config files.
When no data is available in either graph, the graph now just shows text saying "No data available."

MCJack123 commented:
I've gone ahead and made the changes I requested for you, as well as some other tweaks for 1.20/1.21, and I'll be merging it now. Thank you for your contributions!
